### ISSUE TITLE
Incorrect corner collision

### ISSUE THREAD

**Issue by dvdfu** [kikito/bump.lua#33]
Hey, I'm seeing something strange with collision detection that I think is a bug. The scenario looks like this:

![bump](https://cloud.githubusercontent.com/assets/6414395/22498330/84d736c2-e80c-11e6-82e2-b26aa40c4912.png)

- blue rectangle is at (0,32) and is 16x16
- red rectangle is at (16,0) and is 32x32
- size/position are arbitrary, but their corners meet up precisely and do **NOT** overlap
- blue rectangle attempts to move upward and to the left with the `slide` filter

Clearly, there shouldn't be a collision, but after calling `world:move`, there's a collision object returned. I would see `ay = 32` and `col.normal.y = 1`.

I think these results aren't intended.

I guess it's worth mentioning: this was never a problem for me up until I decided to use fixed (not delta-time multiplied) displacement vectors. I tested a bit and realised this would happen whenever the coordinates lined up precisely like so.

**Comment by kikito** [comment]
Hi,

> their corners meet up precisely and do NOT overlap

Given how floating points work in Lua, I had to add a little "delta" or 0.0000000001 on certain operations, in order to mitigate floating-point errors. The practical effect of that is that rectangles in bump are sometimes "surrounded" (so to speak) by a tiny membrane of 0.0000000001 pixels. 

It might be the cause for what you are seeing.

Can you provide a piece of code showing the problem?

> I tested a bit and realised this would happen whenever the coordinates lined up precisely like so

I warn you that corners have been a problem in bump since the beginning. I have tried many times to fix them, but they always come back. It might be that I am implementing things badly, or it might be that the strategy that I used for solving this has inherent defects. If you are using fixed vectors maybe you don't need bump.

**Comment by dvdfu** [comment]
Here's a code snippet - I think it's the most isolated example that produces the bug.

```lua
local bump = require 'bump.bump'

function love.load()
    local world = bump.newWorld(32)
    local A = {}
    local B = {}
    world:add(A, 16, 0, 16, 16)
    world:add(B, 0, 16, 16, 16)

    local ax, ay, cols, len = world:move(B, -1, 15)
    for i = 1, len do
        local col = cols[i]
        print(ax, ay, col.normal.y, col.ti)
    end
end
```

The print output is `-1	16	1	-0`

The game I'm making is a platformer, and I discovered the issue when I used a fixed timestep and wall jumped: some wall jumps actually caused the player to move straight down because they'd collide with the bottom side of a wall, despite being to its left and not touching it at all. I think using imprecise timesteps could eliminate the bug for all practical purposes, but bump has been pretty fundamental to the game's development so far (thank you for that!)

**Comment by kikito** [comment]
I tried something a bit unorthodox - I just added a special case for when a rect "touches the corner" of another one. It doesn't seem to have any ill effects. I have tried it on the main demo and it seems to work just fine.

¿Do you mind trying it a bit in your code, and see whether it works ok? You can find it on the `fix-33` branch:

https://github.com/kikito/bump.lua/blob/fix-33/bump.lua

The change is just this line: https://github.com/kikito/bump.lua/commit/79c32c149acf5bd10cb434327f27a7bae6d02a6e

If it works ok on your end, I will add this to master and include your code (and some mirror cases, moving in other directions etc) as a test.

**Comment by dvdfu** [comment]
Just pulled the code and tested, seemed to work fine for me. Couldn't reproduce my issue after a lot of attempts, or in the isolated attempts.

Thanks so much!

**Comment by kikito** [comment]
Alright then. I have been thinking that maybe that if must be put somewhere else (it's not that function's responsibility to decide when to skip that case). But I will add equivalent conditions somewhere else, add the tests etc, and release a new version as soon as I can. Thank you for helping me with this!

**Comment by kikito** [comment]
I have added a fix and released version 3.1.7 of the library with it.

I want to thank you for taking the time to report the issue, and for your help in finding the problem (The code you provided was of great help) and testing the solution. I'm sorry if I have not been very fast at releasing an "official" solution. It should work just like the one I gave you the other day, but the check is in a more appropiate place.

I will be closing this issue now. Please try the new version and reopen it if the problem reappears.
